### DISCUSSION TITLE
32bit sound on PCM5102A with new librarys.

### DISCUSSION THREAD

**Discussion by Jusis2002** [pschatzmann/ESP32-A2DP#541] [Q&A] [answered]
Code from "bt_music_receiver_32bits.ino" give me strange compilation error for DOITESP32 DEVKIT V1 board ?
Here is final lines from long log:

Using library audio-tools at version 0.9.6 in folder: C:\Users\ccc\Documents\Arduino\libraries\arduino-audio-tools-main 
Using library WiFiClientSecure at version 2.0.0 in folder: C:\Users\ccc\AppData\Local\Arduino15\packages\esp32\hardware\esp32\2.0.14\libraries\WiFiClientSecure 
Using library WiFi at version 2.0.0 in folder: C:\Users\ccc\AppData\Local\Arduino15\packages\esp32\hardware\esp32\2.0.14\libraries\WiFi 
Using library audio-driver at version 0.0.1 in folder: C:\Users\ccc\Documents\Arduino\libraries\arduino-audio-driver-main 
Using library SPI at version 2.0.0 in folder: C:\Users\ccc\AppData\Local\Arduino15\packages\esp32\hardware\esp32\2.0.14\libraries\SPI 
Using library Wire at version 2.0.0 in folder: C:\Users\ccc\AppData\Local\Arduino15\packages\esp32\hardware\esp32\2.0.14\libraries\Wire 
Using library ESP32-A2DP at version 1.7.3 in folder: C:\Users\ccc\Documents\Arduino\libraries\ESP32-A2DP-main 
"C:\\Users\\ccc\\AppData\\Local\\Arduino15\\packages\\esp32\\tools\\xtensa-esp32-elf-gcc\\esp-2021r2-patch5-8.4.0/bin/xtensa-esp32-elf-size" -A "C:\\Users\\ccc\\AppData\\Local\\Temp\\arduino\\sketches\\1051965F0FC4754136DE785B0FB5EDE2/Phil_Schatzmann_new_library.ino.elf"
Sketch uses 1387217 bytes (105%) of program storage space. Maximum is 1310720 bytes.
Global variables use 49684 bytes (15%) of dynamic memory, leaving 277996 bytes for local variables. Maximum is 327680 bytes.
Sketch too big; see https://support.arduino.cc/hc/en-us/articles/360013825179 for tips on reducing it.
text section exceeds available space in board

Compilation error: text section exceeds available space in board

**Comment by pschatzmann** [comment, marked as answer]
[RTFM](https://github.com/pschatzmann/ESP32-A2DP/wiki/Program-Memory)

**Reply by Jusis2002** [reply]
Thanks, all compile.
But where live default  I2S pin config it seams my DAC have other pins ?

**Comment by pschatzmann** [comment]
Just connect the default pins as documented in the Readme or define the pins that you used in the sketch

**Comment by Jusis2002** [comment]
In old library i use trick for prevent PCM5102 DAC from audio "0" detecting with add small static error (pcmLeft+1) :

                    data32[i*2+0] = (pcmLeft+1)  * volumeFactor * 6 ; // ori=*16
                    data32[i*2+1] = (pcmRight+1) * volumeFactor * 6 ; 
where in new lib i manage this ?

**Reply by pschatzmann** [reply]
You can overwrite write_data, like you did before and just do the 32 bit conversion yourself and add this step.
Instead of using i2s_write you would use out->write(data, len)

**Reply by Jusis2002** [reply]
Similar code i see in A2DPVolumeControl.h , but how to use this lib?
Can you give simple example ?

**Comment by pschatzmann** [comment]
I am not sure if your formula makes any sense. But that's your responsibility!

This has not been tested, but to give some directions:

## Alternative 1
Adapting the old example:
```C++
class BluetoothA2DPSink32 : public BluetoothA2DPSink {
    protected:
        void audio_data_callback(const uint8_t *data, uint32_t len) {
            ESP_LOGD(BT_AV_TAG, "%s", __PRETTY_FUNCTION__);
            // update volume
            volume_control()->update_audio_data((Frame*)data, len/4);
            // data is provided as array of int16 
            int sample_count = len / sizeof(int16_t);
            int16_t* data16 = static_cast<int16_t*>(data);
            for (int j=0; j<sample_count; j++){
                 // convert to 32 bit data
                 int32_t sample32 = static_cast<int32_t>(data16[j]) << 16;
                // add your special logic here
                // output 32 bits to DAC
                 out->write((uint8_t*)&sample32, sizeof(sample32);
        }
};

AudioInfo info(44100, 2, 32);
I2SStream out;
BluetoothA2DPSink32 a2dp_sink(out);

void setup() {
  Serial.begin(115200);
  // AudioLogger::instance().begin(Serial, AudioLogger::Info);

  // Configure i2s to use 32 bits
  auto cfg = out.defaultConfig();
  cfg.copyFrom(info);
  out.begin(cfg);

  // start a2dp
  a2dp_sink.start("AudioKit");  
}


void loop() {
  delay(1000); // do nothing
}
```

## Alternative 2
Use the latest AudioTools you can add a formula to a converter

```C++
#include "AudioTools.h"
#include "AudioLibs/AudioBoardStream.h" // install https://github.com/pschatzmann/arduino-audio-driver
#include "BluetoothA2DPSink.h"

AudioInfo info(44100, 2, 32);
I2SStream out;
// just add 1 to the output
CallbackConverterT<int32_t> converter([](int32_t in, int channel) { return in+1; });
ConverterStream cs(out, converter);
NumberFormatConverterStream convert32(cs);
BluetoothA2DPSink a2dp_sink(convert32);

void setup() {
  Serial.begin(115200);
  // AudioLogger::instance().begin(Serial, AudioLogger::Info);

  // Configure i2s to use 32 bits
  auto cfg = out.defaultConfig();
  cfg.copyFrom(info);
  out.begin(cfg);

  // Convert from 16 to 32 bits
  convert.begin(16, 32);

  // start converter stream
  cs.begin();

  // start a2dp
  a2dp_sink.start("AudioKit");  
}


void loop() {
  delay(1000); // do nothing
}
```

## Alternative 3
Just use the callback method 

```C++
#include "BluetoothA2DPSink.h"

AudioInfo info(44100, 2, 32);
I2SStream out;
BluetoothA2DPSink a2dp_sink(out);

// callback 
void get_audio_data(const uint8_t *data, uint32_t len) {
    // data is provided as array of int16 
    int sample_count = len / sizeof(int16_t);
    int16_t* data16 = static_cast<int16_t*>(data);
    for (int j=0; j<sample_count; j++){
         // convert to 32 bit data
         int32_t sample32 = static_cast<int32_t>(data16[j]) << 16;
        // add your special logic here
        // output 32 bits to DAC
         out.write((uint8_t*)&sample32, sizeof(sample32);
}

void setup() {

  // Configure i2s to use 32 bits
  auto cfg = out.defaultConfig();
  cfg.copyFrom(info);
  out.begin(cfg);

  a2dp_source.start_raw("MyMusic", get_audio_data, false);  
}

```

**Reply by Jusis2002** [reply]
This syntax in v2 not compile:
"CallbackConverterT<int32_t> converter([](int32_t in, int channel) { return in+1; }); "

**Reply by pschatzmann** [reply]
Just committed a correction to the AudioTools, but I guess the preferred alternative would be no 3

**Reply by Jusis2002** [reply]
Did you try to compile it yourself?
For me not compile v3, i reinstall AudioTools too.

**Comment by pschatzmann** [comment]
I did not try the whole example, but the line CallbackConverterT<int32_t> converter([](int32_t in, int channel) { return in+1; });  only compiles

**Comment by Jusis2002** [comment]
In A2DPVolumeControl.h is the very similar code like in BluetoothA2DPSink32.h


        virtual void update_audio_data(Frame* data, uint16_t frameCount) {
            if (data!=nullptr && frameCount>0 && ( mono_downmix || is_volume_used)) {
                ESP_LOGD("VolumeControl", "update_audio_data");
                for (int i=0;i<frameCount;i++){
                    int32_t pcmLeft = data[i].channel1;
                    int32_t pcmRight = data[i].channel2;
                    // if mono -> we provide the same output on both channels
                    if (mono_downmix) {
                        pcmRight = pcmLeft = (pcmLeft + pcmRight) / 2;
                    }
                    // adjust the volume
                    if (is_volume_used) {
                        pcmLeft = (pcmLeft+1) * volumeFactor / volumeFactorMax;
                        pcmRight = (pcmRight+1) * volumeFactor / volumeFactorMax;
                    }
                    data[i].channel1 = pcmLeft;
                    data[i].channel2 = pcmRight;
                }
            }
        }

Why do not use A2DPVolumeControl for my needs?
Plaese tell how to better declare it in your new bt_music_receiver_32bits.ino ?

This is very important, otherwise classical music cannot be heard even in 32-bit mode with the PCM5102A!
Because all quiet places are spoiled by the automatic data detection algorithm, which cannot be disabled in this chip.
As a result of the operation of this algorithm, there is a pop in the sound.
I discussed this issue with TI engineers and they said that the only way is to artificially create a static error of 1 bit.

In any way second problem is about volume managing, because min volume from iPHONE is very loud.

**Reply by pschatzmann** [reply]
Because the Volume control is 16 bits and if you add a 1 there the impact is much too big.
Just adapt the callback solution that I was proposing!

if
    int32_t sample32 = static_cast<int32_t>(data16[j]) << 16; 

is too loud, you can adapt it to use your own factor.

**Reply by Jusis2002** [reply]
I would do it but I have a compilation error for V3:

C:\Users\ccc\Documents\Arduino\V3_Phill\V3_Phill.ino: In function 'void get_audio_data(const uint8_t*, uint32_t)':
C:\Users\ccc\Documents\Arduino\V3_Phill\V3_Phill.ino:13:49: error: invalid static_cast from type 'const uint8_t*' {aka 'const unsigned char*'} to type 'int16_t*' {aka 'short int*'}
     int16_t* data16 = static_cast<int16_t*>(data);
                                                 ^
C:\Users\ccc\Documents\Arduino\V3_Phill\V3_Phill.ino: In function 'void setup()':
C:\Users\ccc\Documents\Arduino\V3_Phill\V3_Phill.ino:30:3: error: 'a2dp_source' was not declared in this scope
   a2dp_source.start_raw("MyMusic", get_audio_data, false);
   ^~~~~~~~~~~
C:\Users\ccc\Documents\Arduino\V3_Phill\V3_Phill.ino:30:3: note: suggested alternative: 'a2dp_sink'
   a2dp_source.start_raw("MyMusic", get_audio_data, false);
   ^~~~~~~~~~~
   a2dp_sink

Using library audio-tools at version 0.9.6 in folder: C:\Users\ccc\Documents\Arduino\libraries\arduino-audio-tools-main 
Using library WiFiClientSecure at version 2.0.0 in folder: C:\Users\ccc\AppData\Local\Arduino15\packages\esp32\hardware\esp32\2.0.14\libraries\WiFiClientSecure 
Using library WiFi at version 2.0.0 in folder: C:\Users\ccc\AppData\Local\Arduino15\packages\esp32\hardware\esp32\2.0.14\libraries\WiFi 
Using library audio-driver at version 0.0.1 in folder: C:\Users\ccc\Documents\Arduino\libraries\arduino-audio-driver-main 
Using library SPI at version 2.0.0 in folder: C:\Users\ccc\AppData\Local\Arduino15\packages\esp32\hardware\esp32\2.0.14\libraries\SPI 
Using library Wire at version 2.0.0 in folder: C:\Users\ccc\AppData\Local\Arduino15\packages\esp32\hardware\esp32\2.0.14\libraries\Wire 
Using library ESP32-A2DP at version 1.7.3 in folder: C:\Users\ccc\Documents\Arduino\libraries\ESP32-A2DP-main 
exit status 1

Compilation error: invalid static_cast from type 'const uint8_t*' {aka 'const unsigned char*'} to type 'int16_t*' {aka 'short int*'}

**Reply by pschatzmann** [reply]
Please file an issue, but I think you messed up with the versions!

**Comment by Jusis2002** [comment]
I talk about last v3 code, which is not compile :

#include "AudioTools.h"
#include "AudioLibs/AudioBoardStream.h" // install https://github.com/pschatzmann/arduino-audio-driver
#include "BluetoothA2DPSink.h"

AudioInfo info(44100, 2, 32);
I2SStream out;
BluetoothA2DPSink a2dp_sink(out);

// callback 
void get_audio_data(const uint8_t *data, uint32_t len) {
    // data is provided as array of int16 
    int sample_count = len / sizeof(int16_t);
    int16_t* data16 = static_cast<int16_t*>(data);
    for (int j=0; j<sample_count; j++){
         // convert to 32 bit data
         int32_t sample32 = static_cast<int32_t>(data16[j]) << 2;
        // add your special logic here
        // output 32 bits to DAC
         out.write((uint8_t*)&sample32, sizeof(sample32));
    }
}

void setup() {

  // Configure i2s to use 32 bits
  auto cfg = out.defaultConfig();
  cfg.copyFrom(info);
  out.begin(cfg);

  a2dp_source.start_raw("MyMusic", get_audio_data, false);  
}

**Reply by pschatzmann** [reply]
And I talk about the ESP32-A2DP at version 1.7.3 which you seem to be using. The actual version is 1.8.1!
After all, this discussion is about the new version of the A2DP library.

**Reply by Jusis2002** [reply]
Strange ,  in [library.properties](https://github.com/pschatzmann/ESP32-A2DP/blob/main/library.properties)
i not see version 1.8.1 for ESP32-A2DP , only version 1.7.3?
Where i download actual version 1.8.1 ?

**Reply by pschatzmann** [reply]
OK, I see - I did not update the information in library.properties!

In GIT, versions are managed with the help of tags (and branches).  If you use the main branch, you get the latest version including the __actual corrections__ that make it into the next version. 

If you click on the "main" drop down in the upper left corner, you can click on tags and select a specific version: downloading this will give you the state when it was released. Alternatively you can click on Releases in the right section to download a specific release.

I would recommend to use the main branch and select a specific release only, if you run into problems. 

And finally, a discussion is here for questions and not to report problems, otherwise I am starting to get lost and will not be able to track the open bugs properly.

**Reply by Jusis2002** [reply]
You won't believe it, but this version also doesn't compile.
PC with WINDOWS 11.
Arduino Version 2.3.2 , the latest.
I tried both the latest Arduino Release v2.0.15 based on ESP-IDF v4.4.7 and Arduino Release v2.0.15 based on ESP-IDF v5.1.

**Reply by pschatzmann** [reply]
As I was writing above, I was giving just some untested code out of my head to demonstrate the different alternatives approaches. There might be some small issues that you were supposed to address yourself.

After all, you are supposed to learn something out of this...

**Comment by pschatzmann** [comment]
The following corrected code does not give any compile errors:
```C++
#include "AudioTools.h"
#include "BluetoothA2DPSink.h"

AudioInfo info(44100, 2, 32);
I2SStream out;
BluetoothA2DPSink a2dp_sink(out);

// callback
void get_audio_data(const uint8_t *data, uint32_t len) {
  // data is provided as array of int16
  int sample_count = len / sizeof(int16_t);
  int16_t* data16 = (int16_t*) data;
  for (int j = 0; j < sample_count; j++) {
    // convert to 32 bit data
    int32_t sample32 = static_cast<int32_t>(data16[j]) << 16;
    // add your special logic here
    // output 32 bits to DAC
    out.write((uint8_t*)&sample32, sizeof(sample32));
  }
}

void setup() {

  // Configure i2s to use 32 bits
  auto cfg = out.defaultConfig();
  cfg.copyFrom(info);
  out.begin(cfg);
  // data callback with no output
  a2dp_sink.set_stream_reader(get_audio_data, false);

  a2dp_sink.start("MyMusic");
}
void loop(){}

```

**Reply by Jusis2002** [reply]
BIG thanks , code compiles,  but sound is pulsed and 1 octave higer as wanted !

**Reply by Jusis2002** [reply]
My be incoming or outgoingdata format   is other (MSB,LSB,PHILPS) ?

**Reply by pschatzmann** [reply]
a2dp_sink.set_stream_reader(get_audio_data, false); The false was supposed to deactivate the internal I2S output, but this somehow got lost. 

I committed a correction to correct that problem. Please notice in the sketch I also corrected the  << 2 to << 16!

**Reply by Jusis2002** [reply]
// callback
void get_audio_data(const uint8_t *data, uint32_t len) {
  // data is provided as array of int16
  int sample_count = len / sizeof(int16_t);
  int16_t* data16 = (int16_t*) data;
  for (int j = 0; j < sample_count; j++) {
    // add static error of 1 bit to each 16 bit sample
    int16_t sample_with_error = data16[j] | 0x0001;
    // convert to 32 bit data with sign extension
    int32_t sample32 = static_cast<int32_t>(sample_with_error) << 16; // Sign extension
    // add your special logic here
    // output 32 bits to DAC
    out.write((uint8_t*)&sample32, sizeof(sample32));
  }
}
Here is working MOD.

**Comment by Jusis2002** [comment]
Thank you very much Phil, now the sound from the PCM5102A is excellent, of course as much as the self-noise of the chip allows.
Is it possible to use the exponential volume control from the library A2DPVolumeControl.h in 32 bit mode?

**Reply by pschatzmann** [reply]
Yes. you can select the exponential volume control instead of the default one which is currently used. It just changes the 16 bit data in a different way before you convert them to 32 bits.

**Reply by Jusis2002** [reply]
Would you show me how to correctly call this function from the library in the above script?

**Reply by pschatzmann** [reply]
Just assign it to the a2dp_sink. It gets automatically called!

**Comment by Jusis2002** [comment]
I see it that way :
#include "A2DPVolumeControl.h" 

A2DPSimpleExponentialVolumeControl volumeControl;

uint8_t volumeLevel = 0;

volumeControl.set_volume(volumeLevel);

int32_t volumeFactor = volumeControl.get_volume_factor();

int32_t maxVolumeFactor = volumeControl.get_volume_factor_max();

**Comment by pschatzmann** [comment]
No, these methods are called by the sink. You just follow the [documentation](https://github.com/pschatzmann/ESP32-A2DP/wiki/Volume-Control)!

**Reply by Jusis2002** [reply]
Here is my new code , but he has absolute silent after add from documentation:

#include "AudioTools.h"
#include "BluetoothA2DPSink.h"

AudioInfo info(44100, 2, 32);
I2SStream out;
BluetoothA2DPSink a2dp_sink(out);
A2DPSimpleExponentialVolumeControl vc;


// callback
void get_audio_data(const uint8_t *data, uint32_t len) {
  // data is provided as array of int16
  int sample_count = len / sizeof(int16_t);
  int16_t* data16 = (int16_t*) data;
  for (int j = 0; j < sample_count; j++) {
    //add static error of 1bit to each 16 bit sample
    int16_t sample_with_error = data16[j] | 0x0001;
    // convert to 32 bit data
    int32_t sample32 = static_cast<int32_t>(sample_with_error) << 16;
    // add your special logic here
    // output 32 bits to DAC
    out.write((uint8_t*)&sample32, sizeof(sample32));
  }
}

void setup() {

  a2dp_sink.set_auto_reconnect(true); 
  // Configure i2s to use 32 bits
  auto cfg = out.defaultConfig();
  cfg.pin_bck  = 26; //27;
  cfg.pin_ws   = 25; // 26;
  cfg.pin_data = 19; //25;
  cfg.copyFrom(info);
  out.begin(cfg);
  a2dp_sink.set_volume_control(&vc);

  // data callback with no output
  a2dp_sink.set_stream_reader(get_audio_data, false);

  a2dp_sink.start("MyMusic");
}
void loop(){

}

**Reply by pschatzmann** [reply]
Strange that nobody complained so far. There was a bug in A2DPSimpleExponentialVolumeControl.
I committed a correction, however I did not have the chance yet to test it...

**Reply by Jusis2002** [reply]
All is OK, exp and default vol.contr. from library work , but linear give error :
Compilation error: 'A2DPLinearVolumeControl::A2DPLinearVolumeControl()' is private within this context

**Reply by pschatzmann** [reply]
It should work now as well...

**Reply by Jusis2002** [reply]
Not work :
C:\Users\ccc\Documents\Arduino\V3_Phill\V3_Phill.ino:7:25: error: 'A2DPLinearVolumeControl::A2DPLinearVolumeControl()' is private within this context
 A2DPLinearVolumeControl vc;
                         ^~
In file included from c:\Users\ccc\Documents\Arduino\libraries\ESP32-A2DP\src/BluetoothA2DPCommon.h:59,
                 from c:\Users\ccc\Documents\Arduino\libraries\ESP32-A2DP\src/BluetoothA2DPSink.h:18,
                 from C:\Users\ccc\Documents\Arduino\V3_Phill\V3_Phill.ino:2:
c:\Users\ccc\Documents\Arduino\libraries\ESP32-A2DP\src/A2DPVolumeControl.h:125:9: note: declared private here
         A2DPLinearVolumeControl() {
         ^~~~~~~~~~~~~~~~~~~~~~~

Using library audio-tools at version 0.9.6 in folder: C:\Users\ccc\Documents\Arduino\libraries\arduino-audio-tools-main 
Using library WiFiClientSecure at version 2.0.0 in folder: C:\Users\ccc\AppData\Local\Arduino15\packages\esp32\hardware\esp32\2.0.15\libraries\WiFiClientSecure 
Using library WiFi at version 2.0.0 in folder: C:\Users\ccc\AppData\Local\Arduino15\packages\esp32\hardware\esp32\2.0.15\libraries\WiFi 
Using library ESP32-A2DP at version 1.8.1 in folder: C:\Users\ccc\Documents\Arduino\libraries\ESP32-A2DP 
exit status 1

Compilation error: 'A2DPLinearVolumeControl::A2DPLinearVolumeControl()' is private within this context

**Reply by pschatzmann** [reply]
I guess you just forgot to install the latest version with the correction ?

**Comment by Jusis2002** [comment]
Here is my version of adaptable exponential volume control :

/**
 * @brief  Exponentional volume control 2
 * @author jusis2002 
 * Default coefficients show how to use summ of 1 and 7,5 dagree polinomials,
 * extracted from parabola formula to make adaptable exponential curve.
 */

class A2DPSimpleExponentialVolumeControl2 : public A2DPVolumeControl {
    virtual void set_volume(uint8_t volume) override {
        double volumeFactorFloat = volume;
        // Coefficients for polinome.
        const float A = 0.3456789;    // Here we adjust the curve startup sector from "0" point. 
        const float B = 7.5;          // Here we adjust the curve middle sector.
        const float C = 6.75E-13;     // Here we adjust the curve final point "4096"
        // Modified parabola formula.
        volumeFactorFloat = A * volumeFactorFloat + pow(volumeFactorFloat,B) * C  ;
        volumeFactor = volumeFactorFloat;
        if (volumeFactor > 0xfff) {
            volumeFactor = 0xfff;
        }
    }
};

Here is curves, yellow is  rbruelma, blue is my:

![Curves rbruelma   jusis2002](https://github.com/pschatzmann/ESP32-A2DP/assets/107631484/6858d9da-b84b-410a-8130-b5af51837d69)
